### ISSUE TITLE
Parameters as dependencies for child containers

### ISSUE THREAD

**Issue by LorienMan** [scribd/Weaver#102]
```
class Service {
    // weaver: inputParam <= Int
    // weaver: worker = Worker
    // weaver: worker.scope = .container

    func someMethod() {
        dependencies.worker.doWork()
    }
}

class Worker {
    // weaver: operation = Operation
    // weaver: operation.scope = .transient

    func doWork() {
        let operation = dependencies.operation
        operation.start()
    }
}

class Operation {
    // weaver: inputParam <- Int

    func start() {
        print(dependencies.inputParam)
    }
}
```
I want to pass a root object param as a dependency to it's child. Is it possible to make it work? Or is there another way of doing such a thing?
Thank you for such a great tool, btw

**Comment by indieSoftware** [comment]
Hi @LorienMan 

You can do it that way:

```
class Service {
	let dependencies = ServiceDependencyContainer(inputParam: 42)

	// weaver: inputParam <= Int
	// weaver: worker = Worker
	// weaver: worker.scope = .container

	func someMethod() {
		dependencies.worker(inputParam: 0).doWork()
	}
}

class Worker {
	let dependencies: WorkerDependencyResolver

	// weaver: inputParam <= Int

	// weaver: operation = Operation
	// weaver: operation.scope = .transient

	init(injecting dependencies: WorkerDependencyResolver) {
		self.dependencies = dependencies
	}

	func doWork() {
		let operation = dependencies.operation(inputParam: dependencies.inputParam)
		operation.start()
	}
}

class Operation {
	let dependencies: OperationDependencyResolver

	// weaver: inputParam <= Int

	init(injecting dependencies: OperationDependencyResolver) {
		self.dependencies = dependencies
	}

	func start() {
		print(dependencies.inputParam)
	}
}
```

However, there are currently some flaws as you can read in issue #99

**Comment by LorienMan** [comment]
Thank you for your answer @indieSoftware  ! I'm doing it exactly as you suggested but I don't really want to pass the param all the way down.

Another problem that is close to the first one:
```
class Service {
    // weaver: inputParam <= Int
    // weaver: worker = Worker
    // weaver: worker.scope = .container
    // weaver: someInnerClass = SomeInnerClass
    // weaver: someInnerClass.scope = .container //(this scope does not really matter here)

    func someMethod() {
        dependencies.someInnerClass.someMethod()
    }
}

class Worker {
    // weaver: workerInputParam <= Int

    func callMe() {}
}

class SomeInnerClass {
    // weaver: worker <- Worker

    func someMethod() {
        dependencies.worker.callMe()
    }
}
```
On the level of `Service` I want to instantiate `Worker` using some param (that may come from outside in dependency resolver). All child containers should access exactly this `Worker` instance without using any params. I can't make it work except passing `Worker` instance as a param to children (like `worker <= Worker`)

**Comment by trupin** [comment]
Hi @LorienMan,

As @indieSoftware mentioned, the parameter annotation isn't really fit to be used on dependencies with a scope `.container` because these are built at initialization of their dependency container, and thus, the only way to make it work is to do something similar to what @indieSoftware proposed. 

This is something I am currently thinking about fixing since it appears to be a recurrent issue when using Weaver.

> I want to instantiate Worker on the level of Service using some param (that may come from outside in dependency resolver). All child containers should access exactly this Worker instance without using any params. I can't make it work except passing Worker instance as a param to children (like worker <= Worker)

The fact you are declaring `// weaver: workerInputParam <= Int` automatically generates all accessors to `Worker` taking a param `workerInputParam: Int`. This is in fact the same issue than the first one. I will work on improving this, but there's no way to do this at the moment.

**Comment by trupin** [comment]
JYI @LorienMan and @indieSoftware, this is being addressed by #121 which will soon get merged and released.

**Comment by LorienMan** [comment]
@trupin It's been 9 months since this issue occurred and we've managed to solve it by using custom builder on `Service` level :)

I'd like to mention that we are happy to use this _experimental_ tool in production for all this time ❤️
